File: Cpp/Undercover.cpp
#pragma once
#include<iostream>
#include<string>
#include<vector>
#include"Buyer.cpp"
#include"Police.cpp"
#include"Farmer.cpp"
using namespace std;

class Undercover : public Buyer, public Police {
private:
    string coverIdentity;
    vector<Farmer> targetList;

public:
    // Default constructor
    Undercover() : Buyer(), Police() {
        this->coverIdentity = "";
    }

    // Parameterized constructor
    Undercover(int idBuyer, string buyerName, string buyerAddress, double budget,
               int idPolice, string policeName, string policeAddress, string badgeNumber, string rank,
               string cover)
        : Buyer(idBuyer, buyerName,buyerAddress, budget),
          Police(idPolice, policeName, policeAddress, badgeNumber, rank) {
        this->coverIdentity = cover;
    }

    // ===== Getter methods =====
    string getCoverIdentity() {
        return this->coverIdentity;
    }

    vector<Farmer> getTargetList() {
        return this->targetList;
    }

    // ===== Setter methods =====
    void setCoverIdentity(string cover) {
        this->coverIdentity = cover;
    }

    void setTargetList(vector<Farmer> daftar) {
        this->targetList = daftar;
    }

    // Tambah target (Farmer)
    void addTarget(const Farmer& farmer) {
        this->targetList.push_back(farmer);
    }

    void displayUndercover(int indent) {
        string sp(indent, ' ');
        cout << sp << "Undercover ID       : " << getIdPerson() << endl;
        cout << sp << "Real Name           : " << getName() << endl;
        cout << sp << "Real Address        : " << getAddress() << endl;
        cout << sp << "Budget              : " << getMaxBudget() << endl; // dari Buyer
        cout << sp << "Badge Number        : " << getBadgeNumber() << endl; // dari Police
        cout << sp << "Rank                : " << getRank() << endl;        // dari Police
        cout << sp << "Cover Identity      : " << getCoverIdentity() << endl << endl;

        if (!targetList.empty()) {
            cout << sp << "    Target Farmers:" << endl;
            for (auto &f : targetList) {
                f.displayFarmer(indent + 8); // Tampilkan detail Farmer
            }
        }
        cout << endl;
    }

    // Destructor
    ~Undercover() {
    }
};

File: Python/Undercover.py
from Buyer import Buyer
from Police import Police
from Farmer import Farmer

class Undercover(Buyer, Police):
    def __init__(self, 
                 idBuyer: int = 0, buyerName: str = "", buyerAddress: str = "", budget: float = 0.0,
                 idPolice: int = 0, policeName: str = "", policeAddress: str = "", 
                 badgeNumber: str = "", rank: str = "",
                 cover: str = ""):
        # Inisialisasi parent Buyer
        Buyer.__init__(self, idBuyer, buyerName, buyerAddress, budget)
        # Inisialisasi parent Police
        Police.__init__(self, idPolice, policeName, policeAddress, badgeNumber, rank)

        # Atribut tambahan
        self.coverIdentity = cover
        self.targetList: list[Farmer] = []

    # ===== Getter methods =====
    def getCoverIdentity(self) -> str:
        return self.coverIdentity

    def getTargetList(self) -> list:
        return self.targetList

    # ===== Setter methods =====
    def setCoverIdentity(self, cover: str):
        self.coverIdentity = cover

    def setTargetList(self, daftar: list):
        self.targetList = daftar

    # Tambah target (Farmer)
    def add_Target(self, farmer: Farmer):
        self.targetList.append(farmer)

    # Display method
    def display_undercover(self, indent: int = 0):
        sp = " " * indent
        print(f"{sp}Undercover ID       : {self.get_id_person()}")
        print(f"{sp}Real Name           : {self.get_name()}")
        print(f"{sp}Real Address        : {self.get_address()}")
        print(f"{sp}Budget              : {self.get_max_budget()}")   # dari Buyer
        print(f"{sp}Badge Number        : {self.get_badge_number()}") # dari Police
        print(f"{sp}Rank                : {self.get_rank()}")        # dari Police
        print(f"{sp}Cover Identity      : {self.getCoverIdentity()}\n")

        if self.targetList:
            print(f"{sp}    Target Farmers:")
            for f in self.targetList:
                f.display_farmer(indent + 8)

        print()

